cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/WrapperUtil.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.batch;
+
+import co.cask.cdap.api.ProgramLifecycle;
+import co.cask.cdap.common.lang.ClassLoaders;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.util.ReflectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ */
+public final class WrapperUtil {
+  private static final Logger LOG = LoggerFactory.getLogger(WrapperUtil.class);
+
+  private WrapperUtil() {
+  }
+
+  // instantiates and initializes (if its a ProgramLifeCycle) the class specified by the attribute
+  static <T> T createDelegate(Configuration conf, String attrClass) {
+    String delegateClassName = conf.get(attrClass);
+    Class<?> delegateClass = conf.getClassByNameOrNull(delegateClassName);
+    Preconditions.checkNotNull(delegateClass, "Class could not be found: ", delegateClassName);
+    T delegate = (T) ReflectionUtils.newInstance(delegateClass, conf);
+
+    if (!(delegate instanceof ProgramLifecycle)) {
+      return delegate;
+    }
+
+    MapReduceClassLoader classLoader = MapReduceClassLoader.getFromConfiguration(conf);
+    BasicMapReduceTaskContext basicMapReduceContext = classLoader.getTaskContextProvider().get(conf);
+    ClassLoader programClassLoader = classLoader.getProgramClassLoader();
+
+    ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(programClassLoader);
+    try {
+      ((ProgramLifecycle) delegate).initialize(new MapReduceLifecycleContext(basicMapReduceContext));
```

Comment:
if the delegate is a Configurable, then we need to call its setConf() method, too.